platforms/allwinner-d1/core/src/lib.rs

```diff
@@ -260,14 +265,32 @@ impl D1 {
     }
 
     pub fn handle_panic(info: &PanicInfo) -> ! {
+        // Disable interrupts.
+        unsafe {
+            riscv::interrupt::disable();
+        }
+
+        // Avoid double panics.
+        static PANICKING: AtomicBool = AtomicBool::new(false);
+        if PANICKING.swap(true, Ordering::SeqCst) {
+            die();
+        }
+
+        // TODO(eliza): abort any in-flight DMAs.
+
         // Ugly but works
         let mut uart: Uart = unsafe { core::mem::transmute(()) };
 
         write!(&mut uart, "\r\n").ok();
         write!(&mut uart, "{}\r\n", info).ok();
+        uart.write(&[0]);
 
-        loop {
-            core::sync::atomic::fence(Ordering::SeqCst);
+        die();
+
+        fn die() -> ! {
+            loop {
+                core::sync::atomic::fence(Ordering::SeqCst);
```

Comment:
Not a huge deal, but you could wfi loop here instead to not heat the CPU (it'll probably just never wake)

Reply:
oh good call